examples/search_augmentation.ipynb

```diff
@@ -0,0 +1,472 @@
+{
+ "cells": [
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Search Augmentation\n",
+    "\n",
+    "### with Multiple Query Generation and Semantic Reranking\n",
+    "\n",
+    "Searching for information can be challenging. We can leverage the completions API and embeddings to help us sift through the noise. In this notebook, we will use the completions API to generate search queries given a user's question, and then rerank the results using semantic similarity to a hypothetical answer.\n",
+    "\n",
+    "We can break down this process into three steps:\n",
+    "\n",
+    "**1. Search**\n",
+    "\n",
+    "- User asks a question\n",
+    "- Model generates a list of queries\n",
+    "- Search queries are executed in parallel\n",
+    "\n",
+    "**2. Re-rank**\n",
+    "\n",
+    "- Model generates an ideal answer by hallucination\n",
+    "- Search results are ranked based on semantic similarity to the ideal answer\n",
+    "\n",
+    "**3. Answer**\n",
+    "\n",
+    "- Given the top search results, the model attempts to answer the user question, including references and links.\n",
+    "\n",
+    "Let's dive into it! We will use Twitter as an example domain to search over.\n",
+    "\n",
+    "## Setup\n",
+    "\n",
+    "Once you have your keys, you can set them as environment variables in your`.env` file in the same directory as this notebook. The `.env` file should look like this:\n",
+    "\n",
+    "```\n",
+    "NEWS_API_KEY=your_api_key\n",
+    "OPENAI_API_KEY=your_openai_api_key\n",
+    "```\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Dependencies\n",
+    "import openai\n",
+    "from tqdm import tqdm\n",
+    "import os\n",
+    "import dotenv\n",
+    "import requests\n",
+    "import json\n",
+    "from datetime import date, timedelta\n",
+    "from numpy import dot\n",
+    "from IPython import display\n",
+    "\n",
+    "\n",
+    "# Load environment variables\n",
+    "dotenv.load_dotenv()\n",
+    "\n",
+    "news_api_key = os.getenv(\"NEWS_API_KEY\")\n",
+    "\n",
+    "\n",
+    "# Helper functions\n",
+    "def json_gpt(prompt):\n",
+    "    completion = openai.ChatCompletion.create(\n",
+    "        model=\"gpt-3.5-turbo\",\n",
+    "        messages=[\n",
+    "            {\"role\": \"system\", \"content\": \"Output only valid JSON\"},\n",
+    "            {\"role\": \"user\", \"content\": prompt},\n",
+    "        ],\n",
+    "        temperature=1,\n",
+    "    )\n",
+    "\n",
+    "    text = completion.choices[0].message.content\n",
+    "    parsed = json.loads(text)\n",
+    "\n",
+    "    return parsed\n",
+    "\n",
+    "\n",
+    "def embedding(input):\n",
+    "    response = openai.Embedding.create(\n",
+    "        model=\"text-embedding-ada-002\", input=input)\n",
+    "    return [data.embedding for data in response.data]\n"
```

Comment:
could optionally name the function embeddings rather than embedding, given that it always returns a list (makes me think plural rather than singular)